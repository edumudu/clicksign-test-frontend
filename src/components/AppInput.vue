<template>
  <label
    class="flex flex-col"
    :class="[$style.wrapper, classes]"
    :style="style"
  >
    <span
      v-show="label"
      :class="$style.label"
    >
      {{ label }}
    </span>

    <input
      ref="field"
      v-model="innerValue"
      type="text"
      :class="[$style.field, $style[variant]]"
      v-bind="$attrs"
    >
  </label>
</template>

<script lang="ts">
import {
  defineComponent,
  ref,
  Ref,
  computed,
  SetupContext,
  toRefs,
  onMounted,
  PropType,
} from 'vue';

const useInput = (modelValue: Ref<string | number>, emit: SetupContext['emit']) => {
  const innerValue = computed({
    get: () => modelValue.value,
    set: (value) => emit('update:modelValue', value),
  });

  return {
    innerValue,
  };
};

const useAutoFocus = (autofocus: Ref<boolean>) => {
  const field = ref<HTMLInputElement | null>(null);

  onMounted(() => {
    if (autofocus.value && field.value) field.value.focus();
  });

  return {
    field,
  };
};

export default defineComponent({
  name: 'AppInput',
  inheritAttrs: false,

  props: {
    label: {
      type: String,
      default: '',
    },

    modelValue: {
      type: [String, Number],
      default: '',
    },

    style: {
      type: [String, Object],
      default: '',
    },

    class: {
      type: [String, Object, Array],
      default: '',
    },

    variant: {
      type: String as PropType<'primary' | 'secondary'>,
      default: 'primary',
      validator: (value: string) => ['primary', 'secondary'].includes(value),
    },

    autofocus: Boolean,
  },

  setup(props, { emit }) {
    const { modelValue, autofocus } = toRefs(props);

    return {
      ...useInput(modelValue, emit),
      ...useAutoFocus(autofocus),

      classes: props.class,
    };
  },
});
</script>

<style lang="scss" module>
.wrapper + .wrapper { margin-top: 1rem; }

.label {
  font-size: 0.875rem;
  color: var(--text-color);
  margin-bottom: 0.25rem;
}

.field {
  border: 1px solid var(--gray);
  border-radius: 0.25rem;
  color: var(--text-color);
  font-size: 0.875rem;
  padding: 0.5625rem 0.625rem;
  outline: none;
  font-family: 'Roboto', sans-serif;

  &:focus {
    border-color: var(--color-primary);
  }

  &.secondary {
    background-color: var(--gray-lighter);
    border: none;
  }
}
</style>
