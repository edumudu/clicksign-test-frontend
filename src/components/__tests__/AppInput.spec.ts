import { shallowMount, VueWrapper } from '@vue/test-utils';

import AppInput from '../AppInput.vue';

describe('AppAvatar.vue', () => {
  let wrapper: VueWrapper<any>;

  beforeEach(() => {
    wrapper = shallowMount(AppInput);
  });

  it('should match snapshot', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it('Should render type="text" input as default', () => {
    expect(wrapper.find('input').attributes('type')).toBe('text');
  });

  it.each([
    'email',
    'tel',
    'link',
  ])('Should render type="%s" input as default', async (type) => {
    await wrapper.setProps({ type });

    expect(wrapper.find('input').attributes('type')).toBe(type);
  });

  it('Should be focused when autofocus is true', () => {
    wrapper = shallowMount(AppInput, {
      props: { autofocus: true },
      attachTo: document.body,
    });

    expect(wrapper.find('input').element).toBe(document.activeElement);
  });

  it('Should render label prop text', async () => {
    await wrapper.setProps({ label: 'test' });

    expect(wrapper.text()).toContain('test');
  });
});
