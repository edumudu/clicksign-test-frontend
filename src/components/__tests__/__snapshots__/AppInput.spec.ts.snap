// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppAvatar.vue should match snapshot 1`] = `
<label
  class="flex flex-col wrapper"
>
  <span
    class="label"
    style="display: none;"
  />
  <input
    class="field"
    type="text"
  />
</label>
`;
